openfl-tutorials/interactive_api/PyTorch_BrainMRI_UNet/envoy/shard_descriptor.py

```diff
@@ -0,0 +1,300 @@
+        self.csv = pd.read_csv(csv_path)
+        self.crop_dim = crop_dims
+        self.dataset_dir = dataset_dir
+        # Shard Dataset
+        if set == 'train':
+            if self.rank == self.worldsize:
+                self.csv = self.csv.loc[(self.rank - 1) * (
+                    len(self.csv) // self.worldsize):, :].reset_index(
+                        drop=True)
+            else:
+                self.csv = self.csv.loc[(self.rank - 1) * (len(
+                    self.csv) // self.worldsize):(self.rank) * (len(
+                        self.csv) // self.worldsize), :].reset_index(drop=True)
+        self.device = device
+        self.augment = augment
+
+    def __getitem__(self, idx) -> Tuple[Any, Any]:
+        """Return a item by the index."""
+        img_path = os.path.join(self.dataset_dir, self.csv.loc[idx, 'img_path'])
+        msk_path = os.path.join(self.dataset_dir, self.csv.loc[idx, 'msk_path'])
+        img = np.load(img_path)
+        msk = np.load(msk_path)
+        if self.crop_dim is None:
+            self.crop_dim = [img.shape[0], img.shape[1]]
+        img = self.preprocess_img(img)
+        msk = self.preprocess_label(msk)
+        img, msk = self.crop_input(img, msk)
+        if self.augment:
+            img, msk = self.augment_data(img, msk)
+        img = img[np.newaxis, :, :].copy()
+        msk = msk.copy()
+        img = torch.tensor(img).to(self.device).float()
+        msk = torch.tensor(msk).to(self.device).int()
+        return img, msk
+
+    def __len__(self) -> int:
+        """Return the len of the dataset."""
+        return len(self.csv)
+
+    def crop_input(self, img, msk) -> Tuple[Any, Any]:
+        """
+        Randomly crop the image and mask
+        """
+        # If crop_dim == -1, then don't crop
+        if self.crop_dim[0] == -1:
+            self.crop_dim[0] = img.shape[0]
+        if self.crop_dim[1] == -1:
+            self.crop_dim[1] = img.shape[1]
+
+        slices = []
+
+        # Do we randomize?
+        is_random = self.augment and np.random.rand() > 0.5
+
+        for idx, idy in enumerate(range(2)):  # Go through each dimension
+
+            cropLen = self.crop_dim[idx]
+            imgLen = img.shape[idy]
+
+            start = (imgLen - cropLen) // 2
+
+            ratio_crop = 0.20  # Crop up this this % of pixels for offset
+            # Number of pixels to offset crop in this dimension
+            offset = int(np.floor(start * ratio_crop))
+
+            if offset > 0:
+                if is_random:
+                    start += np.random.choice(range(-offset, offset))
+                    if ((start + cropLen) > imgLen):  # Don't fall off the image
+                        start = (imgLen - cropLen) // 2
+            else:
+                start = 0
+
+            slices.append(slice(start, start + cropLen))
+
+        return img[tuple(slices)], msk[tuple(slices)]
+
+    def augment_data(self, img, msk) -> Tuple[Any, Any]:
```

Comment:
Augmenting data is a part of FL experiment. Shard descriptor just provides local data